receiver/googlecloudmonitoringreceiver/receiver.go

```diff
@@ -5,51 +5,268 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+		}
+
+		// Create an iterator for the time series data
+		tsIter := mr.client.ListTimeSeries(ctx, tsReq)
+		mr.logger.Debug("Retrieving time series data")
+
+		// Iterate over the time series data
+		for {
+			timeSeries, err := tsIter.Next()
+			if errors.Is(err, iterator.Done) {
+				break
+			}
+
+			// Handle errors and break conditions for the iterator
+			if err != nil {
+				gErr = fmt.Errorf("failed to retrieve time series data: %w", err)
+				return metrics, gErr
+			}
+
+			// Convert and append the metric directly within the loop
+			mr.convertGCPTimeSeriesToMetrics(metrics, metricDesc, timeSeries)
+		}
+	}
+
+	return metrics, gErr
+}
+
+// metricDescriptorAPI fetches and processes metric descriptors from the monitoring API.
+func (mr *monitoringReceiver) metricDescriptorAPI(ctx context.Context) error {
+	// Iterate over each metric in the configuration to calculate start/end times and construct the filter query.
+	for _, metric := range mr.config.MetricsList {
+		// Get the filter query for the metric
+		filterQuery := getFilterQuery(metric)
+
+		// Define the request to list metric descriptors
+		metricReq := &monitoringpb.ListMetricDescriptorsRequest{
+			Name:   "projects/" + mr.config.ProjectID,
+			Filter: filterQuery,
+		}
+
+		// Create an iterator for the metric descriptors
+		metricIter := mr.client.ListMetricDescriptors(ctx, metricReq)
+
+		// Iterate over the time series data
+		for {
+			metricDesc, err := metricIter.Next()
+			if errors.Is(err, iterator.Done) {
+				break
+			}
+
+			// Handle errors and break conditions for the iterator
+			if err != nil {
+				return fmt.Errorf("failed to retrieve metric descriptors data: %w", err)
+			}
+			allMetricDescriptor[metricDesc.Type] = metricDesc
+		}
+	}
+
+	mr.logger.Info("Successfully retrieved all metric descriptors.")
 	return nil
 }
+
+// calculateStartEndTime calculates the start and end times based on the current time, interval, and delay.
+func calculateStartEndTime(interval, delay time.Duration) (time.Time, time.Time) {
+	// Get the current time
+	now := time.Now()
+
+	// Calculate end time by subtracting delay
+	endTime := now.Add(-delay)
+
+	// Calculate start time by subtracting interval from end time
+	startTime := endTime.Add(-interval)
+
+	// Return start and end times
+	return startTime, endTime
+}
+
+// getFilterQuery constructs a filter query string based on the provided metric.
+func getFilterQuery(metric MetricConfig) string {
+	var filterQuery string
+	const baseQuery = `metric.type =`
+
+	// If a specific metric name is provided, use it in the filter query
+	filterQuery = fmt.Sprintf(`%s "%s"`, baseQuery, metric.MetricName)
+	return filterQuery
+}
+
+// ConvertGCPTimeSeriesToMetrics converts GCP Monitoring TimeSeries to pmetric.Metrics
+func (mr *monitoringReceiver) convertGCPTimeSeriesToMetrics(metrics pmetric.Metrics, metricDesc *metric.MetricDescriptor, timeSeries *monitoringpb.TimeSeries) {
+	rm := metrics.ResourceMetrics().AppendEmpty()
```

Comment:
This is more correct than it was previously, but still not optimal.  It will now create a new ResourceMetrics for each TimeSeries.  This can result in duplicate resources within a batch of ResourceMetrics if multiple TimeSeries are associated with the same MonitoredResource (i.e. have the same UserLabels and SystemLabels).  Ideally, if this is a resource we have seen before, we would append a new Metric to the existing ResourceMetric.

E.g.
TimeSeries:
  - Resource=foo, Metric=bar
  - Resource=foo, Metric=baz
  - Resource=foo, Metric=dong
  
  currently becomes:
  
resourcemetrics:
- resource: foo
   scopemetrics:
   metrics:
   - name: bar
- resource: foo
   scopemetrics:
   metrics:
   - name: baz
- resource: foo
   scopemetrics:
   metrics:
   - name: dong

We would instead prefer:

resourcemetrics:
- resource: foo
   scopemetrics:
   metrics:
   - name: bar
   - name: baz
   - name: dong